File: python/hardware/hil_write_analog_example/hil_write_analog_example.py
#################################################################
#
# hil_write_analog_example.py - Python file
#
# This example writes one sample immediately to four analog output channels.
#
# This example demonstrates the use of the following functions:
#    HIL.open
#    HIL.write_analog
#    HIL.close
#
# Copyright (C) 2023 Quanser Inc.
#################################################################

from quanser.hardware import HIL, HILError, Clock
import math
import array as arr

board_type = "q2_usb"
board_identifier = "0"

try:
    card = HIL()
    card.open(board_type, board_identifier)

    try:
        channels = arr.array("i", [0, 1])
        num_channels = len(channels)

        voltages = arr.array("d", [0.0] * num_channels)

        for channel in range(num_channels):
            voltages[channel] = (channel - 1.5)

        print("This example writes constant voltages to the first %d analog output" % num_channels)
        print("channels. The voltages written are:")
        for channel in range(num_channels):
            print("    DAC[%d] = %4.1f V" % (channels[channel], voltages[channel]))

        card.write_analog(channels, num_channels, voltages)

    except HILError as ex:
        print("Unable to write channels. %s" % ex.get_error_message())

    finally:
        card.close()

except HILError as ex:
    print("Unable to open board. %s" % ex.get_error_message())

input("Press Enter to continue.")
exit(0)

File: C/hardware/hil_write_analog_example/hil_write_analog_example.c
//////////////////////////////////////////////////////////////////
//
// hil_write_analog_example.c - C file
//
// This example writes one sample immediately to four analog output channels.
//
// This example demonstrates the use of the following functions:
//    hil_open
//    hil_write_analog
//    hil_close
//
// Copyright (C) 2008 Quanser Inc.
//////////////////////////////////////////////////////////////////
    
#include "hil_write_analog_example.h"

int main(int argc, char* argv[])
{
    static const char board_type[]       = "q2_usb";
    static const char board_identifier[] = "0";
    static char       message[512];

    qsigaction_t action;
    t_card board;
    t_int  result;

    /* Prevent Ctrl+C from stopping the application so hil_close gets called */
    action.sa_handler = SIG_IGN;
    action.sa_flags   = 0;
    qsigemptyset(&action.sa_mask);

    qsigaction(SIGINT, &action, NULL);

    result = hil_open(board_type, board_identifier, &board);
    if (result == 0)
    {
        const t_uint32 channels[] = { 0, 1 };

        #define NUM_CHANNELS        ARRAY_LENGTH(channels)

        t_double voltages[NUM_CHANNELS];

        t_uint32 channel;
        for (channel = 0; channel < NUM_CHANNELS; channel++)
            voltages[channel] = (channel - 1.5);

        printf("This example writes constant voltages to the first %u analog output\n", NUM_CHANNELS);
        printf("channels. The voltages written are:\n");
        for (channel = 0; channel < NUM_CHANNELS; channel++)
            printf("    DAC[%d] = %4.1f V\n", channels[channel], voltages[channel]);

        result = hil_write_analog(board, channels, NUM_CHANNELS, voltages);
        if (result < 0)
        {
            msg_get_error_message(NULL, result, message, ARRAY_LENGTH(message));
            printf("Unable to write channels. %s Error %d.\n", message, -result);
        }
            
        hil_close(board);
    }
    else
    {
        msg_get_error_message(NULL, result, message, ARRAY_LENGTH(message));
        printf("Unable to open board. %s Error %d.\n", message, -result);
    }

    printf("\nPress Enter to continue.\n");
    getchar();

    return 0;
}
